mlflow/transformers/signature.py

```diff
@@ -0,0 +1,202 @@
+        inputs=Schema(
+            [ColSpec(DataType.string, name="query"), ColSpec(DataType.string, name="table")]
+        ),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.QuestionAnsweringPipeline: ModelSignature(
+        inputs=Schema(
+            [ColSpec(DataType.string, name="question"), ColSpec(DataType.string, name="context")]
+        ),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.FeatureExtractionPipeline: ModelSignature(
+        inputs=Schema([ColSpec(DataType.string)]),
+        outputs=Schema([TensorSpec(np.dtype("float64"), [-1], "double")]),
+    ),
+}
+
+
+def infer_or_get_default_signature(
+    pipeline, example=None, model_config=None, flavor_config=None
+) -> ModelSignature:
+    """
+    Assigns a default ModelSignature for a given Pipeline type that has pyfunc support. These
+    default signatures should only be generated and assigned when saving a model iff the user
+    has not supplied a signature.
+    For signature inference in some Pipelines that support complex input types, an input example
+    is needed.
+    """
+    if example:
+        try:
+            timeout = MLFLOW_INPUT_EXAMPLE_INFERENCE_TIMEOUT.get()
+            if timeout and not _IS_UNIX:
+                timeout = None
+                _logger.warning(
+                    "On Windows, timeout is not supported for model signature inference. "
+                    "Therefore, the operation is not bound by a timeout and may hang indefinitely. "
+                    "If it hangs, please consider specifying the signature manually."
+                )
+            return _infer_signature_with_prediction(
+                pipeline, example, model_config, flavor_config, timeout
+            )
+        except MlflowTimeoutError:
+            _logger.warning(
+                "Attempted to generate a signature for the saved model but prediction operation "
+                f"timed out after {timeout} seconds. Falling back to the default signature for the "
+                "pipeline. You can specify a signature manually or increase the timeout "
+                f"by setting the environment variable {MLFLOW_INPUT_EXAMPLE_INFERENCE_TIMEOUT}"
+            )
+        except Exception as e:
+            _logger.error(
+                "Attempted to generate a signature for the saved model or pipeline "
+                f"but encountered an error: {e}"
+            )
+            raise
+
+    for pipeline_type, signature in _DEFAULT_SIGNATURE_FOR_PIPELINES.items():
+        if isinstance(pipeline, pipeline_type):
+            return signature
+
+    _logger.warning(
+        "An unsupported Pipeline type was supplied for signature inference. "
+        "Either provide an `input_example` or generate a signature manually "
+        "via `infer_signature` if you would like to have a signature recorded "
+        "in the MLmodel file."
+    )
+
+
+def _infer_signature_with_prediction(
```

Comment:
This function name is a little bit confusing, it sounds like `prediction` is provided. Could we rename it as _infer_signature_with_example?